homeassistant/components/tuya/light.py

```diff
@@ -0,0 +1,401 @@
+"""Support for the Tuya lights."""
+from __future__ import annotations
+
+import json
+import logging
+from typing import Any
+
+from tuya_iot import TuyaDevice, TuyaDeviceManager
+
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS,
+    ATTR_COLOR_TEMP,
+    ATTR_HS_COLOR,
+    COLOR_MODE_BRIGHTNESS,
+    COLOR_MODE_COLOR_TEMP,
+    COLOR_MODE_HS,
+    COLOR_MODE_ONOFF,
+    DOMAIN as DEVICE_DOMAIN,
+    LightEntity,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .base import TuyaHaEntity
+from .const import (
+    DOMAIN,
+    TUYA_DEVICE_MANAGER,
+    TUYA_DISCOVERY_NEW,
+    TUYA_HA_DEVICES,
+    TUYA_HA_TUYA_MAP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+# Light(dj)
+# https://developer.tuya.com/en/docs/iot/f?id=K9i5ql3v98hn3
+DPCODE_SWITCH = "switch_led"
+DPCODE_WORK_MODE = "work_mode"
+DPCODE_BRIGHT_VALUE = "bright_value"
+DPCODE_TEMP_VALUE = "temp_value"
+DPCODE_COLOUR_DATA = "colour_data"
+DPCODE_COLOUR_DATA_V2 = "colour_data_v2"
+DPCODE_LIGHT = "light"
+
+MIREDS_MAX = 500
+MIREDS_MIN = 153
+
+HSV_HA_HUE_MIN = 0
+HSV_HA_HUE_MAX = 360
+HSV_HA_SATURATION_MIN = 0
+HSV_HA_SATURATION_MAX = 100
+
+WORK_MODE_WHITE = "white"
+WORK_MODE_COLOUR = "colour"
+
+TUYA_SUPPORT_TYPE = {
+    "dj",  # Light
+    "dd",  # Light strip
+    "fwl",  # Ambient light
+    "dc",  # Light string
+    "jsq",  # Humidifier's light
+    "xdd",  # Ceiling Light
+    "xxj",  # Diffuser's light
+    "fs",  # Fan
+}
+
+DEFAULT_HSV = {
+    "h": {"min": 1, "scale": 0, "unit": "", "max": 360, "step": 1},
+    "s": {"min": 1, "scale": 0, "unit": "", "max": 255, "step": 1},
+    "v": {"min": 1, "scale": 0, "unit": "", "max": 255, "step": 1},
+}
+
+DEFAULT_HSV_V2 = {
+    "h": {"min": 1, "scale": 0, "unit": "", "max": 360, "step": 1},
+    "s": {"min": 1, "scale": 0, "unit": "", "max": 1000, "step": 1},
+    "v": {"min": 1, "scale": 0, "unit": "", "max": 1000, "step": 1},
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+):
+    """Set up tuya light dynamically through tuya discovery."""
+    _LOGGER.debug("light init")
+
+    hass.data[DOMAIN][entry.entry_id][TUYA_HA_TUYA_MAP][
+        DEVICE_DOMAIN
+    ] = TUYA_SUPPORT_TYPE
+
+    async def async_discover_device(dev_ids):
+        """Discover and add a discovered tuya light."""
+        _LOGGER.debug(f"light add-> {dev_ids}")
+        if not dev_ids:
+            return
+        entities = _setup_entities(hass, entry, dev_ids)
+        async_add_entities(entities)
+
+    async_dispatcher_connect(
```

Comment:
Wrap this in `entry.async_on_unload` so it does not leak if the integration is reloaded.